pkg/operator/cluster/controller.go

```diff
@@ -163,21 +169,44 @@ func (c *ClusterController) onAdd(obj interface{}) {
 	poolController.StartWatch(cluster.Namespace, cluster.stopCh)
 
 	// Start object store CRD watcher
-	objectStoreController := object.NewObjectStoreController(c.context, c.rookImage, cluster.Spec.HostNetwork)
+	objectStoreController := object.NewObjectStoreController(c.context, c.rookImage, cluster.Spec.HostNetwork, cluster.ownerRef)
 	objectStoreController.StartWatch(cluster.Namespace, cluster.stopCh)
 
 	// Start file system CRD watcher
-	fileController := file.NewFilesystemController(c.context, c.rookImage, cluster.Spec.HostNetwork)
+	fileController := file.NewFilesystemController(c.context, c.rookImage, cluster.Spec.HostNetwork, cluster.ownerRef)
 	fileController.StartWatch(cluster.Namespace, cluster.stopCh)
 
 	// Start mon health checker
 	healthChecker := mon.NewHealthChecker(cluster.mons)
 	go healthChecker.Check(cluster.stopCh)
+
+	// add the finalizer to the crd
+	err = c.addFinalizer(clust)
+	if err != nil {
+		logger.Errorf("failed to add finalizer to cluster crd. %+v", err)
+	}
 }
 
 func (c *ClusterController) onUpdate(oldObj, newObj interface{}) {
 	oldClust := oldObj.(*rookalpha.Cluster).DeepCopy()
 	newClust := newObj.(*rookalpha.Cluster).DeepCopy()
+	logger.Infof("update to cluster %s", newClust.Namespace)
+
+	// Check if the cluster is being deleted. This code path is called when a finalizer is specified in the crd.
+	// When a cluster is requested for deletion, K8s will only set the deletion timestamp if there are any finalizers in the list.
+	// K8s will only delete the crd and child resources when the finalizers have been removed from the crd.
+	if newClust.DeletionTimestamp != nil {
+		logger.Infof("cluster %s has a deletion timestamp", newClust.Namespace)
+		err := c.handleDelete(newClust, time.Duration(clusterDeleteRetryInterval)*time.Second)
```

Comment:
isnt the only think you need to do here is just to remove the finalizer? Do we still need to call `c.handleDelete()`? I thought kubernetes will call `onDelete` which will call the `c.handleDelete()` function once the finalizer was removed.

Reply:
We do need to call it. Before we remove the finalizer we want to perform the cleanup operations. `handleDelete` is where all the cleanup operations are done and in the `onDelete` callback we call the `handleDelete` again just in case there wasn't a finalizer

Reply:
ahh gotcha.